### PR TITLE
feat(cli): Add `Deno.mainModule`

### PR DESCRIPTION
**Pull request by Alhadis**
This PR exposes the path/URL of the original script that was executed on the command-line (previously available as `location.pathname`; removed in #5034). It's the moral equivalent of Node's `process.argv[1]` value, except that it's a URL string instead of an ordinary path. This is because `deno run` can execute remote files directly, making it necessary to distinguish between `/foo/bar` and `https://deno.land/foo/bar`.

Be aware that for `deno eval`, `Deno.mainModule` will contain some fabricated location ending in `__$deno$eval.ts`. I would have set this property to `null` in such cases, but this is literally the first time I've ever touched Rust, and my attempts were met with confusing and frustrating errors (I guess 15 minutes isn't enough to wrap one's head around the basics of Rust…)

### CODE REVIEW

File: cli/js/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -1899,6 +1899,9 @@ declare namespace Deno {
    */
   export const args: string[];
 
+  /** The URL of the file that was originally executed from the command-line. */
+  export const scriptUrl: URL;
+
```
Comment (ry) [inline cli/js/lib.deno.ns.d.ts:1904]: Please move this to cli/js/lib.deno.unstable.d.ts

File: cli/js/runtime_main.ts [hunk 2]
```diff
@@ -95,11 +95,12 @@ export function bootstrapMainRuntime(): void {
     }
   });
 
-  const { args, cwd, noColor, pid, repl, unstableFlag } = runtime.start();
+  const { args, cwd, noColor, pid, repl, scriptUrl, unstableFlag } = runtime.start();
 
   Object.defineProperties(denoNs, {
     pid: readOnly(pid),
     noColor: readOnly(noColor),
+    scriptUrl: readOnly(scriptUrl),
```
Comment (ry) [inline cli/js/runtime_main.ts:103]: Could you make this a getter so that we can make a runtime check if --unstable was provided or not?

File: cli/tests/script_url.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+console.log("script_url", Deno.scriptUrl.pathname);
```
Comment (nayeemrmn) [inline cli/tests/script_url.ts:1]: What's the intuition behind the name `scriptUrl`? How about `mainUrl`?

Comment (Alhadis) [reply cli/tests/script_url.ts:1]: 🤷‍♂️  I started out with `scriptPath` before realising Deno can also execute remote files directly, so I replaced the `-Path` suffix with `-Url`. Didn't dwell on the nomenclature any further than that.

Will fix.

**Comment by Alhadis** [comment]
Okay, done. Hopefully I did that correctly.

File: cli/js/lib.deno.ns.d.ts (hunk 1 above)
Comment (Alhadis) [reply cli/js/lib.deno.ns.d.ts:1904]: Oops, I knew I forgot something…

Comment (Alhadis) [reply cli/js/lib.deno.ns.d.ts]: Sorry about the oversight. Done.

**Comment by Alhadis** [comment]
@ry FYI, the tests pass when I run them locally. The CI failure seems to have been caused by a connection timeout.

File: cli/js/lib.deno.unstable.d.ts [hunk 4]
```diff
@@ -1245,4 +1245,7 @@ declare namespace Deno {
    *  Requires `allow-env` permission.
    */
   export function hostname(): string;
+
+  /** **UNSTABLE**: The URL of the file that was originally executed from the command-line. */
+  export const mainUrl: URL;
```
[the remark this answers is not in the document]
Comment (Alhadis) [reply cli/js/lib.deno.unstable.d.ts:1250]: Can't we return a frozen `URL` instead?

[the remark this answers is not in the document]
Comment (nayeemrmn) [reply cli/js/lib.deno.unstable.d.ts:1250]: We shouldn't get into freezing things:
1. This really does count as monkey-patching a built-in, since it overwrites specced (?) property descriptors.
2. It limits how the class has to be implemented for freezing to work.
    - ...particularly, it won't work for `URL`s because the actual state is stored in a separate hidden object, while the public fields just use getters and setters.

I think the `URL` class exists to make manipulating them easier, so I'm not sure a frozen one even makes sense.

Also, I initially suggested this after not immediately finding any web API that has a `URL` field set in this way.

[the remark this answers is not in the document]
Comment (Alhadis) [reply cli/js/lib.deno.unstable.d.ts:1250]: I'll change it to `mainModule` then, for consistency with `state.global_state.main_module`.

[the remark this answers is not in the document]
Comment (Alhadis) [reply cli/js/lib.deno.unstable.d.ts:1250]: Fair points. I've changed it to a string named `mainModule`. 👍

File: cli/ops/runtime.rs [hunk 5]
```diff
@@ -39,6 +41,18 @@ fn op_start(
   })))
 }
 
+fn op_main_url(
+  state: &State,
+  _args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
```
[the remark this answers is not in the document]
Comment (Alhadis) [reply cli/ops/runtime.rs:48]: Done. 👍

File: cli/ops/runtime.rs [hunk 6]
```diff
@@ -39,6 +41,18 @@ fn op_start(
...
+  let main = &state.borrow().main_module.to_string();
+  let main_url = ModuleSpecifier::resolve_url_or_path(&main)?;
+  let main_path = std::env::current_dir().unwrap().join(main_url.to_string());
+  state.check_read_blind(&main_path, "main_url")?;
```
[the remark this answers is not in the document]
Comment (Alhadis) [reply cli/ops/runtime.rs]: Done.

File: cli/ops/runtime.rs [hunk 7]
```diff
@@ -39,6 +41,18 @@ fn op_start(
...
+  Ok(JsonOp::Sync(json!(&main)))
+}
```
[the remark this answers is not in the document]
Comment (Alhadis) [reply cli/ops/runtime.rs:57]: Thanks!

**Comment by ford04** [comment]
@Alhadis would you mind explaining what the difference is to `import.meta.url`?

Put in a script file, these two outputs are equivalent:
```
console.log(Deno.mainModule)
console.log(import.meta.url)
```
Or is it meant to have a more concise wording?

**Comment by kitsonk** [comment]
It isn't when that module imports another module.  They are only equivalent in the main module.

**Comment by ford04** [comment]
Right, thanks! I just had the usual case in mind, that the script entry point triggers all main-file related code actions, so `import.meta.url` would be equivalent here. `Deno.mainModule` certainly is more flexible and a good addition.
